### PR TITLE
NC | NSFS | Schema validation

### PR DESCRIPTION
**Pull request by naveenpaul1**
### Explain the changes
1. Buckets and accounts schema validation
2. creation_date added to bucket management CLI

### Issues: Fixed #xxx / Gap #xxx
1. https://issues.redhat.com/browse/MCGI-204

### Testing Instructions:
1. Test with invalid schema JSON for bucket and accounts, Should throw an error
2. Try to create bucket with an invalid creation date, if the creation date missing current data use

- [ ] Doc added/updated
- [ ] Tests added

### CODE REVIEW

File: src/sdk/bucketspace_fs.js [hunk 1]
```diff
@@ -98,16 +99,41 @@ class BucketSpaceFS extends BucketSpaceSimpleFS {
...
+    async validate_account_schema(account) {
+        if (!account.name || !account.email) {
+            throw new RpcError('MISSING_ACCOUNT_INFO', 'Missing account info.');
+        }
+        if (!account.access_keys[0] || !account.access_keys[0].access_key || !account.access_keys[0].secret_key) {
+            throw new RpcError('INVALID_ACCESS', 'Missing account info.');
+        }
+        if (!account.nsfs_account_config || account.nsfs_account_config.uid === undefined ||
+            account.nsfs_account_config.gid === undefined || !account.nsfs_account_config.new_buckets_path) {
+            throw new RpcError('MISSING_NSFS_ACCOUNT_CONFIGURATION', 'Missing account configuration.');
```
Comment (shirady) [inline src/sdk/bucketspace_fs.js:118]: Why did you add "NSFS" ('MISSING_NSFS_ACCOUNT_CONFIGURATION') and in the other cases you didn't?

File: src/sdk/bucketspace_fs.js [hunk 2]
```diff
@@ -98,16 +99,41 @@ class BucketSpaceFS extends BucketSpaceSimpleFS {
...
             dbg.error('BucketSpaceFS.read_account_by_access_key: failed with error', err);
-            throw new RpcError('NO_SUCH_ACCOUNT', `Account with access_key not found`, err);
+            if (err.code === 'ENOENT') {
+                throw new RpcError('NO_SUCH_ACCOUNT', `Account with access_key not found.`, err);
+            }
+            throw new RpcError('NO_SUCH_ACCOUNT', err.message);
+        }
+    }
+
+    async validate_account_schema(account) {
```
Comment (shirady) [inline src/sdk/bucketspace_fs.js:109]: Why is it an async function?

File: src/sdk/bucketspace_fs.js [hunk 3]
```diff
@@ -98,16 +99,41 @@ class BucketSpaceFS extends BucketSpaceSimpleFS {
...
+        if (!account.access_keys[0] || !account.access_keys[0].access_key || !account.access_keys[0].secret_key) {
+            throw new RpcError('INVALID_ACCESS', 'Missing account info.');
+        }
+        if (!account.nsfs_account_config || account.nsfs_account_config.uid === undefined ||
+            account.nsfs_account_config.gid === undefined || !account.nsfs_account_config.new_buckets_path) {
+            throw new RpcError('MISSING_NSFS_ACCOUNT_CONFIGURATION', 'Missing account configuration.');
         }
     }
 
+    async validate_bucket_schema(bucket) {
```
Comment (shirady) [inline src/sdk/bucketspace_fs.js:122]: Why is it an async function?

File: src/sdk/bucketspace_fs.js [hunk 4]
```diff
@@ -98,16 +99,41 @@ class BucketSpaceFS extends BucketSpaceSimpleFS {
...
+        if (!bucket.name || !bucket.system_owner || !bucket.bucket_owner) {
+            throw new RpcError('INVALID_BUCKET_INFO', 'Missing bucket info.');
+        }
+        if (!bucket.path || !bucket.creation_date) {
+            throw new RpcError('INVALID_BUCKET_STATE', 'Missing bucket storage details.');
```
Comment (shirady) [inline src/sdk/bucketspace_fs.js:127]: If I see an error "'Missing bucket storage details." how can I know how to fix it?

File: src/sdk/bucketspace_fs.js [hunk 5]
```diff
@@ -98,16 +99,41 @@ class BucketSpaceFS extends BucketSpaceSimpleFS {
...
+            throw new RpcError('INVALID_ACCESS', 'Missing account info.');
+        }
+        if (!account.nsfs_account_config || account.nsfs_account_config.uid === undefined ||
+            account.nsfs_account_config.gid === undefined || !account.nsfs_account_config.new_buckets_path) {
+            throw new RpcError('MISSING_NSFS_ACCOUNT_CONFIGURATION', 'Missing account configuration.');
         }
     }
 
+    async validate_bucket_schema(bucket) {
+        if (!bucket.name || !bucket.system_owner || !bucket.bucket_owner) {
```
Comment (shirady) [inline src/sdk/bucketspace_fs.js:123]: What is `system_owner`?
I saw it is by default `new SensitiveString(email)` (but the `bucket_owner` is also defined that way.

File: src/cmd/manage_nsfs.js [hunk 6]
```diff
@@ -148,8 +150,10 @@ async function fetch_bucket_data(argv, config_root, from_file) {
             versioning: 'DISABLED',
             path: bucket_path,
             should_create_underlying_storage: true,
+            creation_date: creation_date,
         };
     }
+    update_bucket_creation_date(data);
```
Comment (shirady) [inline src/cmd/manage_nsfs.js:156]: Why do you set the `creation_date` in the `data` and then update it (`update_bucket_creation_date`)?
You can update it and then set it in the data, the function would return the date instead of changing the property in place, for example:
```suggestion
   creation_date: get_updated_bucket_creation_date(creation_date),
```

File: src/cmd/manage_nsfs.js [hunk 7]
```diff
@@ -172,6 +176,14 @@ async function update_bucket_object(config_root, target) {
     return data;
 }
 
+async function update_bucket_creation_date(data) {
```
Comment (shirady) [inline src/cmd/manage_nsfs.js:179]: Why is it an async function?

**Review by shirady — COMMENTED** [review]
Hi,
Added a couple of comments and questions.

File: src/cmd/manage_nsfs.js [hunk 8]
```diff
@@ -172,6 +176,14 @@ async function update_bucket_object(config_root, target) {
...
+    if (data.creation_date) {
+        data.creation_date = Date.parse(data.creation_date) ? new Date(Date.parse(data.creation_date)).toISOString() : '';
+    } else {
+        data.creation_date = new Date().toISOString();
```
[the remark this answers is not in the document]
Comment (naveenpaul1) [reply src/cmd/manage_nsfs.js:183]: `toISOString()` will format the date

File: src/cmd/manage_nsfs.js [hunk 9]
```diff
@@ -61,6 +61,7 @@ Bucket Options:
     # required for add, replace 
     --email <email>             (default none)          Set the email for account/bucket.
     --path <dir>                (default none)          Set the bucket path.
+    --creation_date <time>      (default current time)          Set time bucket creation. Formate YYYY-MM-DDTHH:mm:ssZ,
```
[the remark this answers is not in the document]
Comment (naveenpaul1) [reply src/cmd/manage_nsfs.js:64]: creation date is mandatory for S3 bucket, S3 bucket format is `YYYY-MM-DDTHH:mm:ssZ`, user will get  the format from --help argument

File: src/sdk/bucketspace_fs.js [hunk 10]
```diff
@@ -98,16 +99,41 @@ class BucketSpaceFS extends BucketSpaceSimpleFS {
...
+            throw new RpcError('NO_SUCH_ACCOUNT', err.message);
+        }
+    }
+
+    async validate_account_schema(account) {
+        if (!account.name || !account.email) {
+            throw new RpcError('MISSING_ACCOUNT_INFO', 'Missing account info.');
+        }
+        if (!account.access_keys[0] || !account.access_keys[0].access_key || !account.access_keys[0].secret_key) {
+            throw new RpcError('INVALID_ACCESS', 'Missing account info.');
```
[the remark this answers is not in the document]
Comment (naveenpaul1) [reply src/sdk/bucketspace_fs.js:114]: access and secret key is a mix of alphanumeric characters with special characters not sure very specific validation is required here.

File: src/sdk/bucketspace_fs.js (hunk 1 above)
Comment (naveenpaul1) [reply src/sdk/bucketspace_fs.js:118]: this check is specific to NSFS, normal account will not have account configuration.

File: src/sdk/bucketspace_fs.js (hunk 5 above)
Comment (naveenpaul1) [reply src/sdk/bucketspace_fs.js:123]: a person creating the bucket, in our case both remain the same.

File: src/cmd/manage_nsfs.js (hunk 6 above)
Comment (naveenpaul1) [reply src/cmd/manage_nsfs.js:156]: we can't do that because we need to do the same logic for --from-file flow too.

**Review by shirady — APPROVED** [review]
LGTM

@naveenpaul1, please check if the validation that that we have a value (without checking something about the value) is enough to meet the requirement.

File: src/cmd/manage_nsfs.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (shirady) [reply src/cmd/manage_nsfs.js:183]: As understood, the user would not need to use milliseconds, so it should be fine (probably a minute will be enough).

File: src/sdk/bucketspace_fs.js [hunk 11]
```diff
@@ -98,16 +99,41 @@ class BucketSpaceFS extends BucketSpaceSimpleFS {
...
             dbg.error('BucketSpaceFS.read_account_by_access_key: failed with error', err);
-            throw new RpcError('NO_SUCH_ACCOUNT', `Account with access_key not found`, err);
+            if (err.code === 'ENOENT') {
+                throw new RpcError('NO_SUCH_ACCOUNT', `Account with access_key not found.`, err);
+            }
+            throw new RpcError('NO_SUCH_ACCOUNT', err.message);
+        }
+    }
+
+    validate_account_schema(account) {
```
Comment (romayalon) [inline src/sdk/bucketspace_fs.js:109]: hmm I think we want _ajv validation and not a custom validation, while creating a schema, validating etc, @liranmauda @guymguym WDYT ?

Comment (romayalon) [reply src/sdk/bucketspace_fs.js:109]: same for the bucket validation

**Review by romayalon — CHANGES_REQUESTED** [review]

File: src/sdk/bucketspace_fs.js (hunk 11 above)
Comment (romayalon) [reply src/sdk/bucketspace_fs.js:109]: adding here a reference for ajv validation we do in postgres client - https://github.com/noobaa/noobaa-core/blob/master/src/util/postgres_client.js#L1322

File: src/cmd/manage_nsfs.js (hunk 9 above)
Comment (romayalon) [inline src/cmd/manage_nsfs.js:64]: can you please update the bucket schema of the nsfs standalone doc with this, also I wonder if we want to have also creation_date on account as well, WDYT?

Comment (naveenpaul1) [reply src/cmd/manage_nsfs.js:64]: so the `creation_date` will be current date, right?

[the remark this answers is not in the document]
Comment (naveenpaul1) [reply src/cmd/manage_nsfs.js:64]: if the creation_date is empty then the current date is used, So you suggest removing the `creation_date` flag, is that right?

[the remark this answers is not in the document]
Comment (romayalon) [reply src/cmd/manage_nsfs.js:64]: I'm open to the option to have this flag, I just think that people will prefer to have it always created by us, no?
@PravinRanjan10 do you think the CLI should have a creation_date flag?

**Review by romayalon — COMMENTED** [review]
I think we should also use the schema validation in bucketspace_fs on create bucket :)

File: src/cmd/manage_nsfs.js (hunk 9 above)
Comment (romayalon) [reply src/cmd/manage_nsfs.js:64]: yeah, sounds good to me

File: src/cmd/manage_nsfs.js [hunk 12]
```diff
@@ -61,6 +61,7 @@ Bucket Options:
     # required for add, replace 
     --email <email>             (default none)          Set the email for account/bucket.
     --path <dir>                (default none)          Set the bucket path.
+    --creation_date <time>      (default current time)  Set time bucket creation. Formate YYYY-MM-DDTHH:mm:ssZ,
```
Comment (liranmauda) [inline src/cmd/manage_nsfs.js:64]: Why do we need the creation_date to be exposed as a flag?

**Review by liranmauda — CHANGES_REQUESTED** [review]
@naveenpaul1 
few comments/questions:
1. Am I correct and the creation and read of the account/bucket are not calling any RPC? 
2. please change the names of those Schema, to reflect that those are standalone (or any other name) schemas and not a RPC ones. Also, we already have files with those names so it can get us confused.

File: src/cmd/manage_nsfs.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (liranmauda) [reply src/cmd/manage_nsfs.js]: @romayalon I just commented the same without seeing this comment... 
@naveenpaul1 I also think that we should not have this flag

[the remark this answers is not in the document]
Comment (PravinRanjan10) [reply src/cmd/manage_nsfs.js]: > I'm open to the option of having this flag, I just think that people will prefer to have it always created by us, no? @PravinRanjan10 do you think the CLI should have a creation_date flag?

@romayalon CLI should not have this flag. The internal logic should generate, once bucket creation is successful.

[the remark this answers is not in the document]
Comment (naveenpaul1) [reply src/cmd/manage_nsfs.js]: Will remove the `creation_date`

File: src/cmd/nsfs.js [hunk 13]
```diff
@@ -317,7 +317,7 @@ async function main(argv = minimist(process.argv.slice(2))) {
             }
         });
         if (await endpoint.is_http_allowed(nsfs_config_root)) {
-            dbg.log0('nsfs: listening on', util.inspect(`http://localhost:${http_port}`));
+            console.log('nsfs: listening on', util.inspect(`http://localhost:${http_port}`));
```
Comment (liranmauda) [inline src/cmd/nsfs.js:320]: I wonder, why are we using console and not dbg? 
by requireing the console we "hijak" the console, but console does not have levels like 0,1 ...

**Review by liranmauda — APPROVED** [review]
LGTM

File: src/cmd/nsfs.js (hunk 13 above)
Comment (naveenpaul1) [reply src/cmd/nsfs.js:320]: @liranmauda ya totally agree with you, we need to have a discussion on this. In this PR I just reverted the previous code that I accidentally updated in my last PR.

**Review by romayalon — APPROVED** [review]
